clarity/src/vm/analysis/type_checker/tests/contracts.rs

```diff
@@ -630,3 +664,1864 @@ fn test_expects() {
+            (internal-get-1 contract))
+        (define-public (internal-get-1 (contract <trait-1>))
+            (contract-call? contract get-1 u1))";
+
+        let bad_other_trait = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-get-2 (contract <trait-1>))
+            (internal-get-2 contract))
+        (define-public (internal-get-2 (contract <trait-2>))
+            (contract-call? contract get-2 u1))";
+
+        let embedded_trait = "(define-trait trait-12 (
+            (get-1 (uint) (response uint uint))
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-opt-get-1 (contract <trait-12>))
+            (internal-get-1 (some contract)))
+        (define-public (internal-get-1 (opt-contract (optional <trait-12>)))
+            (match opt-contract
+                contract (contract-call? contract get-1 u1)
+                (err u1)
+            )
+        )";
+
+        let embedded_trait_compatible = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-12 (
+            (get-1 (uint) (response uint uint))
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-opt-get-1 (contract <trait-12>))
+            (internal-get-1 (some contract)))
+        (define-public (internal-get-1 (opt-contract (optional <trait-1>)))
+            (match opt-contract
+                contract (contract-call? contract get-1 u1)
+                (err u1)
+            )
+        )";
+
+        let bad_embedded_trait = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-12 (
+            (get-1 (uint) (response uint uint))
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-opt-get-1 (contract <trait-1>))
+            ;; Passing (optional <trait-1>) as an (optional <trait-12>) should be an error
+            (wrapped-get-1 (some contract)))
+        (define-public (wrapped-get-1 (opt-contract (optional <trait-12>)))
+            (internal-get-1 opt-contract)
+        )
+        (define-public (internal-get-1 (opt-contract (optional <trait-12>)))
+            (match opt-contract
+                contract (contract-call? contract get-1 u1)
+                (err u1)
+            )
+        )";
+
+        let let_trait = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-public (let-echo (t <trait-1>))
+            (let ((t1 t))
+                (contract-call? t1 echo u42)
+            )
+        )";
+
+        let let3_trait = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-public (let-echo (t <trait-1>))
```

Comment:
I like the idea behind this test.  But can you also try other permutations?  Like, can I do something like this?

```clarity
(let ((t1 t))
   (let ((t2-opt (some t1)))
      (let ((t3-res (ok t2-opt))
         (contract-call? t3-res echo u42)
       )
    )
 )
```

Reply:
Added a similar example in 7641da42a. This exact example won't work because `t3-res` is going to be a trait wrapped in an optional wrapped in a response, so we can't do a contract call with it, but we can verify that it is the expected type. This actually works in Clarity1.

```clarity
    (define-private (foo (a (response (optional <trait-1>) uint)))
        (ok true)
    )
    (define-public (let-echo (t <trait-1>))
        (let ((t1 t))
            (let ((t2-opt (some t1)))
                (let ((t3-res (ok t2-opt)))
                    (foo t3-res)
                )
            )
        )
    )
```

We could also then unwrap the trait and use it for a contract call (that would not work in Clarity1):
```clarity
    (define-public (let-echo (t <trait-1>))
        (let ((t1 t))
            (let ((t2-opt (some t1)))
                (let ((t3-res (ok t2-opt)))
                    (let ((t4 (unwrap! (unwrap! t3-res (err u1)) (err u2))))
                        (contract-call? t4 echo u23)
                    )
                )
            )
        )
    )
```